Implement v1 protocol ready handshake

euank opened the pull request:
See the documentation in doc.go and the comment in `transfer_owner.go`'s  `readyHandshake` method for why this protocol change is necessary and what failure modes it has.

This is intended to be a backwards compatible change.

TODO:
- [x] Add a test that uses the previous master of tableroll to do an upgrade handoff with the new one successfully

euank committed:
internal/proto: add versioned json helpers + docs

This adds some documentation for the v1 protocol as well as an
implementation of a hack to encode versions in our json payload that did
not include a version field.

euank committed:
Implement v1 protocol ready handshake

Also fix not closing `conn.File()`'s return in a couple places, which
fixes file descriptor leaks for those connections.

euank commented:
So, the full context what this fixes:

I observed a pair of processes such that both ended up in a state where they thought they were owners. This happened after an upgrade timeout error.

The reason it was possible to get into this state, I believe, is because new process was able to write the ready byte to the connection successfully without the owner ever reading it. This happened very easily on timeout for two reasons:

1. These are not dgram sockets, so it's not totally blocking 1-1 read/writes
2. The parent tried to close the connection, but because it had called `file = conn.File()` on the connection object without a paired `file.Close()`, the connection had been leaked and the child could thus still write to it even after the parent called `conn.Close()`.

Those two facts combined together to make this a very easy bug to hit when a timeout occurred. This change fixes the second bug, and tries to resolve this error case in the first by adding an 'Ack' response and attempting to make the failure scenarios around that ack result in 0 owners rather than 2.

inconshreveable commented on internal/proto/encode.go:
```diff
@@ -0,0 +1,89 @@
+	}
+	if jsonBlobLenBuf.Len() != 4 {
+		panic(fmt.Errorf("int32 should be 4 bytes, not: %+v", jsonBlobLenBuf))
+	}
```
this check seems unnecessary. it's basically a check on the go stdlib, which i don't think we need? we could just binary.Write directly to `dst`

euank replied:
Yeah, you're right, I'll delete it.

inconshreveable reviewed (approved):
lgtm. the whitespace version encoding is clever and horrifying

euank committed:
Add a v0 -> v1 proto upgrade test

This also switches back from conn.SetDeadline to using the clock more
heavily now that I understand that I just need to close the file as well
for that to work.

euank commented on upgrader.go:
```diff
@@ -172,12 +177,11 @@ func (u *Upgrader) handleUpgradeRequest(conn *net.UnixConn) {
 	u.Fds.lockMutations(ErrUpgradeInProgress)
 	// time to pass our FDs along
-	nextOwner, errC := passFdsToSibling(u.l, conn, u.Fds.copy())
+	conn.SetDeadline(u.clock.Now().Add(u.upgradeTimeout))
```
The deadline change was made when I was still trying to figure out why `conn.Close()` didn't work (aka while I was still figuring out I'd leaked a file descriptor).

I can get rid of this in favor of the timer again, which is good because the timer can be mocked, and setdeadline can't be.

euank commented on upgrader_test.go:
```diff
@@ -225,6 +225,7 @@ func TestPIDReuse(t *testing.T) {
 // the same listening fds to multiple processes if a 'Ready' is not received in
 // time.
 func TestFdPassMultipleTimes(t *testing.T) {
+	t.Skip("TODO")
```
This needs to be fixed; it fails because it depends on the mock clock working, and the clock stuff was partly removed.

euank committed:
Remove overtly defensive check

euank commented on go.mod:
```diff
@@ -1,14 +1,12 @@
 require (
-	github.com/go-stack/stack v1.8.0 // indirect
-	github.com/hashicorp/errwrap v1.0.0 // indirect
+	github.com/euank/tableroll v1.0.1-0.20190509205925-1c7855724129
```
This is an absolute hack.

We could also have a `third_party/tableroll/v1_0_0` folder where we vendor our old selves I guess. I'm not sure the best solution to regression testing within a single major version in go.

euank committed:
internal/proto: fix typos

euank committed:
Minor fixups

betawaffle commented on internal/proto/consts.go:
```diff
@@ -0,0 +1,16 @@
+	// Version is the latest version of the protocol. It is implicitly 0 for
+	// clients that didn't yet have a protocol version
+	Version = 1
+	// V0NotifyReady is the value sent at the end in the v0 protocol to indicate
```
If this wasn't so nice and clean, I wouldn't have said anything, but you're missing a blank line between the previous value and the next comment.

betawaffle commented on internal/proto/doc.go:
```diff
@@ -0,0 +1,36 @@
+// The v1 protocol exists because the v0 protocol allows for a new process to
+// think it had notified the previous owner it was ready, even if the new owner
+// never read that byte.
+// The primary reaon this is possible is because the protocol is over a stream
```
`reaon` -> `reason`

betawaffle commented on internal/proto/doc.go:
```diff
@@ -0,0 +1,36 @@
+// return an error from 'Ready()' and exit, and will then restart and get
+// succeed since there will now be no owner.
+// In the case of 2, O must assume that the message sent successfully and
+// actually step down. This again leaves us with 0 owners in the failure mode,
```
`step` -> `stepped`

betawaffle replied:
`0 owners` -> `zero owners`

euank replied:
0->zero definitely helps with readability. I think `step` is what I want here because it's prescribing the behavior that must happen a layer up, not describing something that already happened.

betawaffle replied:
Yeah, I misread it, sorry.

euank committed:
Fix typo

betawaffle commented on internal/proto/encode_version.go:
```diff
@@ -0,0 +1,71 @@
+
+import "fmt"
+
+// This file implements a way to encode and decode a uint32 as json-ignorable
```
Very creative.

betawaffle commented on internal/proto/encode.go:
```diff
@@ -0,0 +1,86 @@
+	var jsonBlob bytes.Buffer
+	jsonBlob.Write(versionPrefix)
+	enc := json.NewEncoder(&jsonBlob)
+	if err := enc.Encode(obj); err != nil {
```
I'd usually just do it inline: `json.NewEncoder(...).Encode(...)`.

betawaffle reviewed (approved):
Seems reasonable, I only found typos. Not sure how well I can actually review the logic, so hopefully your automated and manual testing are good enough to be happy. I trust you in that regard.

euank committed:
Fix typos

euank committed:
Fix test flake in TestUpgradeTimeout

euank commented:
Thanks CI for finding that test flake. It turns out `upg2.Ready` was racing against `conn.Close` + `file.Close` getting called to properly timeout the upgrade. After adding that small sleep as a hack, `go test -count=100` for that test seems happy.
